src/plugins/message-parser/MessageParser.js

```diff
@@ -0,0 +1,86 @@
+import { html } from 'lit-html';
+import { pipe } from '@converse/headless/utils/functional';
+import RichMessage from './RichMessage';
+import log from '@converse/headless/log';
+
+class MessageParser {
+
+    constructor () {
+        this._filters = [];
+        this._middleware = [];
+    }
+
+    get filters () {
+        return this._filters;
+    }
+
+    get middleware () {
+        return this._middleware;
+    }
+
+    static checkReferenceCollision (reference, rich_message) {
+        const collides = rich_message.references.some(msg_ref => {
+          const is_after = reference.begin > msg_ref.end;
+          const is_before = reference.end < msg_ref.begin;
+          return !is_after && !is_before;
+        });
+        return collides;
+    }
+
+    static makeTemplateResultFromString (string_or_templateresult) {
+        const is_template_result = !!string_or_templateresult.getTemplateElement;
+        return is_template_result
+          ? string_or_templateresult
+          : html`${string_or_templateresult}`;
+    }
+
+    static addSingleReferenceToMessage (reference, rich_message) {
+        if (!MessageParser.checkReferenceCollision(reference, rich_message)) {
+            const { begin, end, template } = reference;
+            const template_result = MessageParser.makeTemplateResultFromString(template);
+            rich_message.addTemplateResult(begin, end, template_result);
+        } else {
+            throw new Error('Reference collision');
+        }
+    }
+
+    static addReferencesToMessage (references, rich_message) {
+        references.forEach(reference =>
+            MessageParser.addSingleReferenceToMessage(reference, rich_message)
+        );
+    }
+
+    addFilters (...filters) {
+        this._filters.push(...filters);
+    }
+
+    addMiddleware (...middleware) {
+        this._middleware.push(...middleware);
+    }
+
+    richMessageFromText (text) {
+        return new RichMessage(text);
+    }
+
+    runFilters (text) {
+        return pipe(...this.filters)(text);
+    }
```

Comment:
We cannot have filters that remove parts of the original message, because that would break all the  indexes in the references we have.

The original message text must be considered immutable and read-only.

That's why I created the `MessageText` class in the first place, so that we can save metadata (as `TemplateResult` instances in the `references` attribute) on the `MessageText` instance without having to mutate the original text.

Instead of having filters, I think we can achieve the same result by simply sticking to the current approach whereby we add references that replace the parts that would have been filtered out with `TemplateResult` instances that render nothing (or an empty Text node).